File: examples/scalar_udf/java/app/src/main/java/scalar_udf/SecretsResolver.java
/**
 * Copyright (c) 2026 Jeffrey Jonathan Jennings
 *
 * @author Jeffrey Jonathan Jennings (J3)
 */
package scalar_udf;

import java.net.URI;
import java.util.Optional;

import software.amazon.awssdk.http.urlconnection.UrlConnectionHttpClient;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.secretsmanager.SecretsManagerClient;
import software.amazon.awssdk.services.secretsmanager.SecretsManagerClientBuilder;
import software.amazon.awssdk.services.secretsmanager.model.GetSecretValueRequest;

/**
 * Resolves a UDF secret from one of two sources, in order:
 *
 * <ol>
 *   <li><b>AWS Secrets Manager</b> — when {@code <BASE_NAME>_SECRET_ID} is set, fetch the
 *       secret string for that secret-id. Honors {@code AWS_ENDPOINT_URL_SECRETSMANAGER}
 *       (or the generic {@code AWS_ENDPOINT_URL}) so a LocalStack endpoint can be used in
 *       development and minikube without changing UDF code.</li>
 *   <li><b>Plain environment variable</b> — when {@code <BASE_NAME>_SECRET_ID} is unset,
 *       fall back to {@code <BASE_NAME>_SECRET}. This keeps the UDF runnable in unit tests
 *       and outside Kubernetes.</li>
 * </ol>
 *
 * <p>Picking Secrets Manager is opt-in (driven by the {@code _SECRET_ID} env var) so that
 * pipelines without AWS connectivity continue to work and so that the AWS SDK is exercised
 * only on pods that truly need it.
 */
final class SecretsResolver {

    private SecretsResolver() {}

    /**
     * Resolve the secret for a UDF.
     *
     * @param baseName the env-var base, e.g. {@code "PII_PSEUDONYM"}. The resolver looks at
     *                 {@code baseName + "_SECRET_ID"} (Secrets Manager source) and
     *                 {@code baseName + "_SECRET"} (plain-env fallback).
     * @return the resolved secret string; never {@code null} or empty.
     * @throws IllegalStateException if neither source produces a non-empty value.
     */
    static String resolve(String baseName) {
        String secretId = System.getenv(baseName + "_SECRET_ID");
        if (secretId != null && !secretId.isEmpty()) {
            return fetchFromSecretsManager(secretId, baseName);
        }

        String envSecret = System.getenv(baseName + "_SECRET");
        if (envSecret != null && !envSecret.isEmpty()) {
            return envSecret;
        }

        throw new IllegalStateException(
            "Neither " + baseName + "_SECRET_ID (AWS Secrets Manager source) nor "
            + baseName + "_SECRET (plain environment fallback) is set. Configure one so the "
            + "UDF can load its keying material.");
    }

    private static String fetchFromSecretsManager(String secretId, String baseName) {
        SecretsManagerClientBuilder builder = SecretsManagerClient.builder()
            .httpClient(UrlConnectionHttpClient.create());

        Optional<String> region = firstNonEmpty(
            System.getenv("AWS_REGION"),
            System.getenv("AWS_DEFAULT_REGION"));
        region.ifPresent(r -> builder.region(Region.of(r)));

        Optional<String> endpoint = firstNonEmpty(
            System.getenv("AWS_ENDPOINT_URL_SECRETSMANAGER"),
            System.getenv("AWS_ENDPOINT_URL"));
        endpoint.ifPresent(e -> builder.endpointOverride(URI.create(e)));

        try (SecretsManagerClient client = builder.build()) {
            String value = client.getSecretValue(
                GetSecretValueRequest.builder().secretId(secretId).build()
            ).secretString();
            if (value == null || value.isEmpty()) {
                throw new IllegalStateException(
                    "Secret '" + secretId + "' (referenced by " + baseName + "_SECRET_ID) "
                    + "exists but has an empty SecretString.");
            }
            return value;
        }
    }

    private static Optional<String> firstNonEmpty(String... values) {
        for (String v : values) {
            if (v != null && !v.isEmpty()) return Optional.of(v);
        }
        return Optional.empty();
    }
}

File: examples/scalar_udf/python/secrets_resolver.py
"""
Copyright (c) 2026 Jeffrey Jonathan Jennings

Author: Jeffrey Jonathan Jennings (J3)

Resolves a UDF secret from one of two sources, in order:

1. **AWS Secrets Manager** — when ``<BASE_NAME>_SECRET_ID`` is set, fetch the
   secret string for that secret-id. Honors ``AWS_ENDPOINT_URL_SECRETSMANAGER``
   (or the generic ``AWS_ENDPOINT_URL``) so a LocalStack endpoint can be used
   in development and minikube without changing UDF code.

2. **Plain environment variable** — when ``<BASE_NAME>_SECRET_ID`` is unset,
   fall back to ``<BASE_NAME>_SECRET``. This keeps the UDF runnable in unit
   tests and outside Kubernetes.

Picking Secrets Manager is opt-in (driven by the ``_SECRET_ID`` env var) so
that pipelines without AWS connectivity continue to work and so that boto3 is
imported only on workers that truly need it.
"""
import os


def resolve(base_name: str) -> str:
    """
    Resolve the secret for a UDF.

    :param base_name: env-var base, e.g. ``"PII_PSEUDONYM"``. The resolver
        looks at ``base_name + "_SECRET_ID"`` (Secrets Manager source) and
        ``base_name + "_SECRET"`` (plain-env fallback).
    :return: the resolved secret string; never empty.
    :raises RuntimeError: if neither source produces a non-empty value.
    """
    secret_id = os.environ.get(f"{base_name}_SECRET_ID")
    if secret_id:
        return _fetch_from_secrets_manager(secret_id, base_name)

    env_secret = os.environ.get(f"{base_name}_SECRET")
    if env_secret:
        return env_secret

    raise RuntimeError(
        f"Neither {base_name}_SECRET_ID (AWS Secrets Manager source) nor "
        f"{base_name}_SECRET (plain environment fallback) is set. Configure "
        "one so the UDF can load its keying material."
    )


def _fetch_from_secrets_manager(secret_id: str, base_name: str) -> str:
    # boto3 is imported lazily so jobs that only use the env-var fallback do
    # not pay its import cost (and do not need it installed in the venv).
    import boto3

    kwargs = {}
    region = os.environ.get("AWS_REGION") or os.environ.get("AWS_DEFAULT_REGION")
    if region:
        kwargs["region_name"] = region

    endpoint = (
        os.environ.get("AWS_ENDPOINT_URL_SECRETSMANAGER")
        or os.environ.get("AWS_ENDPOINT_URL")
    )
    if endpoint:
        kwargs["endpoint_url"] = endpoint

    client = boto3.client("secretsmanager", **kwargs)
    response = client.get_secret_value(SecretId=secret_id)
    value = response.get("SecretString")
    if not value:
        raise RuntimeError(
            f"Secret '{secret_id}' (referenced by {base_name}_SECRET_ID) "
            "exists but has an empty SecretString."
        )
    return value
